packages/editor/src/components/provider/use-auto-switch-editor-sidebars.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useEffect } from '@wordpress/element';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { store as preferencesStore } from '@wordpress/preferences';
+import { store as interfaceStore } from '@wordpress/interface';
+
+/**
+ * This listener hook monitors for block selection and triggers the appropriate
+ * sidebar state.
+ */
+function useAutoSwitchEditorSidebars() {
+	const { hasBlockSelection } = useSelect( ( select ) => {
+		return {
+			hasBlockSelection:
+				!! select( blockEditorStore ).getBlockSelectionStart(),
+		};
+	}, [] );
+
+	const { getActiveComplementaryArea } = useSelect( interfaceStore );
+	const { enableComplementaryArea } = useDispatch( interfaceStore );
+	const { get: getPreference } = useSelect( preferencesStore );
```

Comment:
Nit: I personally prefer using the registry now, and call select or dispatch within the hook. It slightly reduces the work on render and reduces the amount of dependencies passed to the effect hook.

Reply:
Interesting :) ... I'm not sure yet.

Do you think the impact is measurable? Because DevX wise, I think "registry" is too low level and few people should know about it.

Reply:
In a lot of cases you need registry anyway to batch. Not sure if it's measurable, I haven't checked